core/src/main/java/org/elasticsearch/discovery/zen/PublishClusterStateAction.java

```diff
@@ -85,21 +88,24 @@
     private final DiscoverySettings discoverySettings;
     private final ClusterName clusterName;
     private final PendingClusterStatesQueue pendingStatesQueue;
+    private final CustomPrototypeRegistry registry;
 
     public PublishClusterStateAction(
             Settings settings,
             TransportService transportService,
             Supplier<ClusterState> clusterStateSupplier,
             NewPendingClusterStateListener listener,
             DiscoverySettings discoverySettings,
-            ClusterName clusterName) {
+            ClusterName clusterName,
+            CustomPrototypeRegistry registry) {
```

Comment:
I might be missing something - but do we need both the NamedWriteable in the stream and the explicit registry when reading a diff? `   Diff<ClusterState> diff = lastSeenClusterState.readDiffFrom(in, registry);`